Application crash on Android at startup

We are sometimes seeing an exception at startup using CodePush, then configuration is basic and we are just using .sync() to deal with everything at device ready.

This was seen when deploying a new version of the app via VS-TACO after already having an earlier build of the app on the device which had already updated itself successfully via CodePush.

The earlier build used exactly the same plugins and had the same binary version, all the changes were in the '.\www' folder. 

This is using CodePush CLI v1.11.0-beta and Cordova CLI 6.0.0

The trace is as follows:

D/AndroidRuntime(13342): Shutting down VM

E/AndroidRuntime(13342): FATAL EXCEPTION: main

E/AndroidRuntime(13342): Process: com.emrgroup.testapp, PID: 13342

E/AndroidRuntime(13342): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.emrgroup.testapp/com.emrgroup.testapp.MainActivity}: java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.microsoft.cordova.CodePushPackageMetadata.label' on a null object reference

E/AndroidRuntime(13342): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2808)

E/AndroidRuntime(13342): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2873)

E/AndroidRuntime(13342): 	at android.app.ActivityThread.access$900(ActivityThread.java:181)

E/AndroidRuntime(13342): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1482)

E/AndroidRuntime(13342): 	at android.os.Handler.dispatchMessage(Handler.java:102)

E/AndroidRuntime(13342): 	at android.os.Looper.loop(Looper.java:145)

E/AndroidRuntime(13342): 	at android.app.ActivityThread.main(ActivityThread.java:6145)

E/AndroidRuntime(13342): 	at java.lang.reflect.Method.invoke(Native Method)

E/AndroidRuntime(13342): 	at java.lang.reflect.Method.invoke(Method.java:372)

E/AndroidRuntime(13342): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)

E/AndroidRuntime(13342): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)

E/AndroidRuntime(13342): Caused by: java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.microsoft.cordova.CodePushPackageMetadata.label' on a null object reference

E/AndroidRuntime(13342): 	at com.microsoft.cordova.CodePush.handleUnconfirmedInstall(CodePush.java:333)

E/AndroidRuntime(13342): 	at com.microsoft.cordova.CodePush.onStart(CodePush.java:446)

E/AndroidRuntime(13342): 	at org.apache.cordova.PluginManager.onStart(PluginManager.java:276)

E/AndroidRuntime(13342): 	at org.apache.cordova.CordovaWebViewImpl.handleStart(CordovaWebViewImpl.java:462)

E/AndroidRuntime(13342): 	at org.apache.cordova.CordovaActivity.onStart(CordovaActivity.java:293)

E/AndroidRuntime(13342): 	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1234)

E/AndroidRuntime(13342): 	at android.app.Activity.performStart(Activity.java:6412)

E/AndroidRuntime(13342): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2771)

E/AndroidRuntime(13342): 	... 10 more

--- a/bin/www/codePush.js
+++ b/bin/www/codePush.js
@@ -25,6 +25,11 @@ var CodePush = (function () {
         cordova.exec(installSuccess, errorCallback, "CodePush", "restartApplication", []);
     };
     CodePush.prototype.reportStatus = function (status, label, appVersion, deploymentKey, previousLabelOrAppVersion, previousDeploymentKey) {
+        if ((!label || label === previousLabelOrAppVersion) &&
+            appVersion === previousLabelOrAppVersion &&
+            deploymentKey == previousDeploymentKey) {
+            return;
+        }
         var createPackageForReporting = function (label, appVersion) {
             return {
                 label: label, appVersion: appVersion, deploymentKey: deploymentKey,

--- a/bin/www/codePush.js
+++ b/bin/www/codePush.js
@@ -27,7 +27,7 @@ var CodePush = (function () {
     CodePush.prototype.reportStatus = function (status, label, appVersion, deploymentKey, previousLabelOrAppVersion, previousDeploymentKey) {
         if ((!label || label === previousLabelOrAppVersion) &&
             appVersion === previousLabelOrAppVersion &&
-            deploymentKey == previousDeploymentKey) {
+            deploymentKey === previousDeploymentKey) {
             return;
         }
         var createPackageForReporting = function (label, appVersion) {

--- a/bin/www/codePush.js
+++ b/bin/www/codePush.js
@@ -25,9 +25,8 @@ var CodePush = (function () {
         cordova.exec(installSuccess, errorCallback, "CodePush", "restartApplication", []);
     };
     CodePush.prototype.reportStatus = function (status, label, appVersion, deploymentKey, previousLabelOrAppVersion, previousDeploymentKey) {
-        if ((!label || label === previousLabelOrAppVersion) &&
-            appVersion === previousLabelOrAppVersion &&
-            deploymentKey === previousDeploymentKey) {
+        if (((!label && appVersion === previousLabelOrAppVersion) || label === previousLabelOrAppVersion)
+            && deploymentKey === previousDeploymentKey) {
             return;
         }
         var createPackageForReporting = function (label, appVersion) {

--- a/src/android/CodePush.java
+++ b/src/android/CodePush.java
@@ -278,7 +278,7 @@ public class CodePush extends CordovaPlugin {
             long nativeBuildTime = Utilities.getApkEntryBuildTime(RESOURCES_BUNDLE, this.cordova.getActivity());
             if (nativeBuildTime != -1) {
                 String currentAppTimeStamp = String.valueOf(nativeBuildTime);
-                if (deployedPackageTimeStamp != null && !deployedPackageTimeStamp.equals(currentAppTimeStamp)) {
+                if (!currentAppTimeStamp.equals(deployedPackageTimeStamp)) {
                     this.codePushPackageManager.cleanDeployments();
                     this.codePushPackageManager.clearFailedUpdates();
                     this.codePushPackageManager.clearPendingInstall();
@@ -298,7 +298,7 @@ public class CodePush extends CordovaPlugin {
                 try {
                     navigateToFile(startPage);
                 } catch (MalformedURLException e) {
-                /* empty - if there is an exception, the app will launch with the bundled content */
+                    /* empty - if there is an exception, the app will launch with the bundled content */
                 }
             }
         }

--- a/src/android/CodePush.java
+++ b/src/android/CodePush.java
@@ -278,14 +278,12 @@ public class CodePush extends CordovaPlugin {
             long nativeBuildTime = Utilities.getApkEntryBuildTime(RESOURCES_BUNDLE, this.cordova.getActivity());
             if (nativeBuildTime != -1) {
                 String currentAppTimeStamp = String.valueOf(nativeBuildTime);
-                if ((deployedPackageTimeStamp != null) && (currentAppTimeStamp != null)) {
-                    if (!deployedPackageTimeStamp.equals(currentAppTimeStamp)) {
-                        this.codePushPackageManager.cleanDeployments();
-                        this.codePushPackageManager.clearFailedUpdates();
-                        this.codePushPackageManager.clearPendingInstall();
-                        this.codePushPackageManager.clearInstallNeedsConfirmation();
-                        this.codePushPackageManager.clearBinaryFirstRunFlag();
-                    }
+                if (deployedPackageTimeStamp != null && !deployedPackageTimeStamp.equals(currentAppTimeStamp)) {
+                    this.codePushPackageManager.cleanDeployments();
+                    this.codePushPackageManager.clearFailedUpdates();
+                    this.codePushPackageManager.clearPendingInstall();
+                    this.codePushPackageManager.clearInstallNeedsConfirmation();
+                    this.codePushPackageManager.clearBinaryFirstRunFlag();
                 }
             }
         }
@@ -293,16 +291,14 @@ public class CodePush extends CordovaPlugin {
 
     private void navigateToLocalDeploymentIfExists() {
         CodePushPackageMetadata deployedPackageMetadata = this.codePushPackageManager.getCurrentPackageMetadata();
-        if (deployedPackageMetadata != null) {
-            if (deployedPackageMetadata.localPath != null) {
-                File startPage = this.getStartPageForPackage(deployedPackageMetadata.localPath);
-                if (startPage != null) {
-                    /* file exists */
-                    try {
-                        navigateToFile(startPage);
-                    } catch (MalformedURLException e) {
-                    /* empty - if there is an exception, the app will launch with the bundled content */
-                    }
+        if (deployedPackageMetadata != null && deployedPackageMetadata.localPath != null) {
+            File startPage = this.getStartPageForPackage(deployedPackageMetadata.localPath);
+            if (startPage != null) {
+                /* file exists */
+                try {
+                    navigateToFile(startPage);
+                } catch (MalformedURLException e) {
+                /* empty - if there is an exception, the app will launch with the bundled content */
                 }
             }
         }

--- a/src/android/CodePush.java
+++ b/src/android/CodePush.java
@@ -109,9 +109,9 @@ public class CodePush extends CordovaPlugin {
     }
 
     private boolean execNotifyApplicationReady(CallbackContext callbackContext) {
-        if (this.codePushPackageManager.isFirstRun()) {
+        if (this.codePushPackageManager.isBinaryFirstRun()) {
             // Report first run of a store version app
-            this.codePushPackageManager.saveFirstRunFlag();
+            this.codePushPackageManager.saveBinaryFirstRunFlag();
             try {
                 String appVersion = Utilities.getAppVersionName(cordova.getActivity());
                 codePushReportingManager.reportStatus(new StatusReport(ReportingStatus.STORE_VERSION, null, appVersion, mainWebView.getPreferences().getString(DEPLOYMENT_KEY_PREFERENCE, null)), this.mainWebView);
@@ -270,6 +270,44 @@ public class CodePush extends CordovaPlugin {
         }
     }
 
+    private void clearDeploymentsIfBinaryUpdated() {
+        /* check if we have a deployed package already */
+        CodePushPackageMetadata deployedPackageMetadata = this.codePushPackageManager.getCurrentPackageMetadata();
+        if (deployedPackageMetadata != null) {
+            String deployedPackageTimeStamp = deployedPackageMetadata.nativeBuildTime;
+            long nativeBuildTime = Utilities.getApkEntryBuildTime(RESOURCES_BUNDLE, this.cordova.getActivity());
+            if (nativeBuildTime != -1) {
+                String currentAppTimeStamp = String.valueOf(nativeBuildTime);
+                if ((deployedPackageTimeStamp != null) && (currentAppTimeStamp != null)) {
+                    if (!deployedPackageTimeStamp.equals(currentAppTimeStamp)) {
+                        this.codePushPackageManager.cleanDeployments();
+                        this.codePushPackageManager.clearFailedUpdates();
+                        this.codePushPackageManager.clearPendingInstall();
+                        this.codePushPackageManager.clearInstallNeedsConfirmation();
+                        this.codePushPackageManager.clearBinaryFirstRunFlag();
+                    }
+                }
+            }
+        }
+    }
+
+    private void navigateToLocalDeploymentIfExists() {
+        CodePushPackageMetadata deployedPackageMetadata = this.codePushPackageManager.getCurrentPackageMetadata();
+        if (deployedPackageMetadata != null) {
+            if (deployedPackageMetadata.localPath != null) {
+                File startPage = this.getStartPageForPackage(deployedPackageMetadata.localPath);
+                if (startPage != null) {
+                    /* file exists */
+                    try {
+                        navigateToFile(startPage);
+                    } catch (MalformedURLException e) {
+                    /* empty - if there is an exception, the app will launch with the bundled content */
+                    }
+                }
+            }
+        }
+    }
+
     private boolean execPreInstall(CordovaArgs args, CallbackContext callbackContext) {
     /* check if package is valid */
         try {
@@ -314,47 +352,6 @@ public class CodePush extends CordovaPlugin {
             callbackContext.success(result);
         } else {
             callbackContext.error("Could not get preference: " + preferenceName);
-        }
-    }
-
-    private void handleAppStart() {
-        try {
-            /* check if we have a deployed package already */
-            CodePushPackageMetadata deployedPackageMetadata = this.codePushPackageManager.getCurrentPackageMetadata();
-            if (deployedPackageMetadata != null) {
-                String deployedPackageTimeStamp = deployedPackageMetadata.nativeBuildTime;
-                long nativeBuildTime = Utilities.getApkEntryBuildTime(RESOURCES_BUNDLE, this.cordova.getActivity());
-                if (nativeBuildTime != -1) {
-                    String currentAppTimeStamp = String.valueOf(nativeBuildTime);
-                    if ((deployedPackageTimeStamp != null) && (currentAppTimeStamp != null)) {
-                        if (deployedPackageTimeStamp.equals(currentAppTimeStamp)) {
-                            /* same native version, safe to launch from local storage */
-                            if (deployedPackageMetadata.localPath != null) {
-                                File startPage = this.getStartPageForPackage(deployedPackageMetadata.localPath);
-                                if (startPage != null) {
-                                    /* file exists */
-                                    navigateToFile(startPage);
-                                }
-                            }
-                        } else {
-                            /* application updated in the store or via local deployment */
-                            this.codePushPackageManager.cleanDeployments();
-                            this.codePushPackageManager.clearFailedUpdates();
-                            this.codePushPackageManager.clearPendingInstall();
-                            this.codePushPackageManager.clearInstallNeedsConfirmation();
-                            try {
-                                String appVersion = Utilities.getAppVersionName(cordova.getActivity());
-                                codePushReportingManager.reportStatus(new StatusReport(ReportingStatus.STORE_VERSION, null, appVersion, mainWebView.getPreferences().getString(DEPLOYMENT_KEY_PREFERENCE, null)), this.mainWebView);
-                            } catch (PackageManager.NameNotFoundException e) {
-                                // Should not happen unless the appVersion is not specified, in which case we can't report anything anyway.
-                                e.printStackTrace();
-                            }
-                        }
-                    }
-                }
-            }
-        } catch (Exception e) {
-            /* empty - if there is an exception, the app will launch with the bundled content */
         }
     }
 
@@ -468,6 +465,7 @@ public class CodePush extends CordovaPlugin {
      */
     @Override
     public void onStart() {
+        clearDeploymentsIfBinaryUpdated();
         if (!didStartApp) {
             /* The application was just started. */
             didStartApp = true;
@@ -478,7 +476,7 @@ public class CodePush extends CordovaPlugin {
                 handleUnconfirmedInstall(false);
             }
 
-            handleAppStart();
+            navigateToLocalDeploymentIfExists();
             /* Handle ON_NEXT_RESUME and ON_NEXT_RESTART pending installations */
             if (pendingInstall != null && (InstallMode.ON_NEXT_RESUME.equals(pendingInstall.installMode) || InstallMode.ON_NEXT_RESTART.equals(pendingInstall.installMode))) {
                 this.markUpdate();
@@ -490,7 +488,7 @@ public class CodePush extends CordovaPlugin {
             InstallOptions pendingInstall = this.codePushPackageManager.getPendingInstall();
             long durationInBackground = (new Date().getTime() - lastPausedTimeMs) / 1000;
             if (pendingInstall != null && InstallMode.ON_NEXT_RESUME.equals(pendingInstall.installMode) && durationInBackground >= pendingInstall.minimumBackgroundDuration) {
-                handleAppStart();
+                navigateToLocalDeploymentIfExists();
                 this.markUpdate();
                 this.codePushPackageManager.clearPendingInstall();
             } else if (codePushReportingManager.hasFailedReport()) {

--- a/src/android/CodePushPackageManager.java
+++ b/src/android/CodePushPackageManager.java
@@ -116,11 +116,15 @@ public class CodePushPackageManager {
         return this.codePushPreferences.installNeedsConfirmation();
     }
 
-    public boolean isFirstRun() {
-        return this.codePushPreferences.isFirstRun();
+    public boolean isBinaryFirstRun() {
+        return this.codePushPreferences.isBinaryFirstRun();
     }
 
-    public void saveFirstRunFlag() {
-        this.codePushPreferences.saveFirstRunFlag();
+    public void clearBinaryFirstRunFlag() {
+        this.codePushPreferences.clearBinaryFirstRunFlag();
+    }
+
+    public void saveBinaryFirstRunFlag() {
+        this.codePushPreferences.saveBinaryFirstRunFlag();
     }
 }

--- a/src/android/CodePushPreferences.java
+++ b/src/android/CodePushPreferences.java
@@ -118,14 +118,21 @@ public class CodePushPreferences {
         return notConfirmedInstall;
     }
 
-    public void saveFirstRunFlag() {
+    public void clearBinaryFirstRunFlag() {
+        SharedPreferences preferences = context.getSharedPreferences(CodePushPreferences.FIRST_RUN_PREFERENCE, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = preferences.edit();
+        editor.remove(CodePushPreferences.FIRST_RUN_PREFERENCE_KEY);
+        editor.commit();
+    }
+
+    public void saveBinaryFirstRunFlag() {
         SharedPreferences preferences = context.getSharedPreferences(CodePushPreferences.FIRST_RUN_PREFERENCE, Context.MODE_PRIVATE);
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean(CodePushPreferences.FIRST_RUN_PREFERENCE_KEY, false);
         editor.commit();
     }
 
-    public boolean isFirstRun() {
+    public boolean isBinaryFirstRun() {
         SharedPreferences preferences = context.getSharedPreferences(CodePushPreferences.FIRST_RUN_PREFERENCE, Context.MODE_PRIVATE);
         boolean isFirstRun = preferences.getBoolean(CodePushPreferences.FIRST_RUN_PREFERENCE_KEY, true);
         return isFirstRun;

--- a/src/ios/CodePush.m
+++ b/src/ios/CodePush.m
@@ -62,9 +62,9 @@ StatusReport* rollbackStatusReport = nil;
 }
 
 - (void)notifyApplicationReady:(CDVInvokedUrlCommand *)command {
-    if ([CodePushPackageManager isFirstRun]) {
+    if ([CodePushPackageManager isBinaryFirstRun]) {
         // Report first run of a store version app
-        [CodePushPackageManager markFirstRunFlag];
+        [CodePushPackageManager markBinaryFirstRunFlag];
         NSString* appVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
         NSString* deploymentKey = ((CDVViewController *)self.viewController).settings[DeploymentKeyPreference];
         StatusReport* statusReport = [[StatusReport alloc] initWithStatus:STORE_VERSION
@@ -92,7 +92,7 @@ StatusReport* rollbackStatusReport = nil;
         [CodePushReportingManager reportStatus:[CodePushReportingManager getAndClearFailedReport]
                                    withWebView:self.webView];
     }
-    
+
     // Mark the update as confirmed and not requiring a rollback
     [CodePushPackageManager clearInstallNeedsConfirmation];
     [CodePushPackageManager cleanOldPackage];
@@ -224,7 +224,7 @@ StatusReport* rollbackStatusReport = nil;
     [[NSNotificationCenter defaultCenter] removeObserver:self];
 }
 
-- (void)handleAppStart {
+- (void)clearDeploymentsIfBinaryUpdated {
     // check if we have a deployed package
     CodePushPackageMetadata* deployedPackageMetadata = [CodePushPackageManager getCurrentPackageMetadata];
     if (deployedPackageMetadata) {
@@ -232,37 +232,37 @@ StatusReport* rollbackStatusReport = nil;
         NSString* applicationBuildTime = [Utilities getApplicationTimestamp];
 
         if (deployedPackageNativeBuildTime != nil && applicationBuildTime != nil) {
-            if ([deployedPackageNativeBuildTime isEqualToString: applicationBuildTime] ) {
-                // same version, safe to launch from local storage
-                if (deployedPackageMetadata.localPath) {
-                    [self redirectStartPageToURL: deployedPackageMetadata.localPath];
-                }
-            }
-            else {
-                // installed native version is different from package version => clean up deployed package and do not modify start page
+            if (![deployedPackageNativeBuildTime isEqualToString: applicationBuildTime]) {
+                // installed native version is different from package version
                 [CodePushPackageManager cleanDeployments];
                 [CodePushPackageManager clearFailedUpdates];
                 [CodePushPackageManager clearPendingInstall];
                 [CodePushPackageManager clearInstallNeedsConfirmation];
-                NSString* appVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
-                NSString* deploymentKey = ((CDVViewController *)self.viewController).settings[DeploymentKeyPreference];
-                StatusReport* statusReport = [[StatusReport alloc] initWithStatus:STORE_VERSION andLabel:nil andAppVersion:appVersion andDeploymentKey:deploymentKey];
-                [CodePushReportingManager reportStatus:statusReport withWebView:self.webView];
+                [CodePushPackageManager clearBinaryFirstRunFlag];
             }
         }
     }
 }
 
+- (void)navigateToLocalDeploymentIfExists {
+    // check if we have a deployed package
+    CodePushPackageMetadata* deployedPackageMetadata = [CodePushPackageManager getCurrentPackageMetadata];
+    if (deployedPackageMetadata && deployedPackageMetadata.localPath) {
+        [self redirectStartPageToURL: deployedPackageMetadata.localPath];
+    }
+}
+
 - (void)pluginInitialize {
     // register for "on resume", "on pause" notifications
+    [self clearDeploymentsIfBinaryUpdated];
     [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationWillEnterForeground) name:UIApplicationWillEnterForegroundNotification object:nil];
     [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationWillResignActive) name:UIApplicationWillResignActiveNotification object:nil];
     InstallOptions* pendingInstall = [CodePushPackageManager getPendingInstall];
     if (!pendingInstall) {
         [self handleUnconfirmedInstall:NO];
     }
-    [self handleAppStart];
 
+    [self navigateToLocalDeploymentIfExists];
     // handle both ON_NEXT_RESUME and ON_NEXT_RESTART - the application might have been killed after transitioning to the background
     if (pendingInstall && (pendingInstall.installMode == ON_NEXT_RESTART || pendingInstall.installMode == ON_NEXT_RESUME)) {
         [self markUpdate];

--- a/src/ios/CodePush.m
+++ b/src/ios/CodePush.m
@@ -233,7 +233,7 @@ StatusReport* rollbackStatusReport = nil;
 
         if (deployedPackageNativeBuildTime != nil && applicationBuildTime != nil) {
             if (![deployedPackageNativeBuildTime isEqualToString: applicationBuildTime]) {
-                // installed native version is different from package version
+                // package version is incompatible with installed native version
                 [CodePushPackageManager cleanDeployments];
                 [CodePushPackageManager clearFailedUpdates];
                 [CodePushPackageManager clearPendingInstall];
@@ -245,7 +245,6 @@ StatusReport* rollbackStatusReport = nil;
 }
 
 - (void)navigateToLocalDeploymentIfExists {
-    // check if we have a deployed package
     CodePushPackageMetadata* deployedPackageMetadata = [CodePushPackageManager getCurrentPackageMetadata];
     if (deployedPackageMetadata && deployedPackageMetadata.localPath) {
         [self redirectStartPageToURL: deployedPackageMetadata.localPath];

--- a/src/ios/CodePushPackageManager.h
+++ b/src/ios/CodePushPackageManager.h
@@ -17,7 +17,8 @@
 + (void)markInstallNeedsConfirmation;
 + (BOOL)installNeedsConfirmation;
 + (void)clearInstallNeedsConfirmation;
-+ (void)markFirstRunFlag;
-+ (BOOL)isFirstRun;
++ (void)clearBinaryFirstRunFlag;
++ (void)markBinaryFirstRunFlag;
++ (BOOL)isBinaryFirstRun;
 
 @end

--- a/src/ios/CodePushPackageManager.m
+++ b/src/ios/CodePushPackageManager.m
@@ -8,7 +8,7 @@ NSString* const BinaryHashKey = @"BINARY_HASH";
 NSString* const FailedUpdatesKey = @"FAILED_UPDATES";
 NSString* const PendingInstallKey = @"PENDING_INSTALL";
 NSString* const NotConfirmedInstallKey = @"NOT_CONFIRMED_INSTALL";
-NSString* const IsFirstRunKey = @"FIRST_RUN";
+NSString* const IsBinaryFirstRunKey = @"FIRST_RUN";
 NSString* const OldPackageManifestName = @"oldPackage.json";
 NSString* const CurrentPackageManifestName = @"currentPackage.json";
 
@@ -120,15 +120,21 @@ NSString* const CurrentPackageManifestName = @"currentPackage.json";
     [preferences removeObjectForKey:PendingInstallKey];
 }
 
-+ (void)markFirstRunFlag {
++ (void)clearBinaryFirstRunFlag {
     NSUserDefaults* preferences = [NSUserDefaults standardUserDefaults];
-    [preferences setBool:YES forKey:IsFirstRunKey];
+    [preferences removeObjectForKey:IsBinaryFirstRunKey];
     [preferences synchronize];
 }
 
-+ (BOOL)isFirstRun {
++ (void)markBinaryFirstRunFlag {
     NSUserDefaults* preferences = [NSUserDefaults standardUserDefaults];
-    BOOL firstRunFlagSet = [preferences boolForKey:IsFirstRunKey];
+    [preferences setBool:YES forKey:IsBinaryFirstRunKey];
+    [preferences synchronize];
+}
+
++ (BOOL)isBinaryFirstRun {
+    NSUserDefaults* preferences = [NSUserDefaults standardUserDefaults];
+    BOOL firstRunFlagSet = [preferences boolForKey:IsBinaryFirstRunKey];
     return !firstRunFlagSet;
 }
 

--- a/www/codePush.ts
+++ b/www/codePush.ts
@@ -65,7 +65,7 @@ class CodePush implements CodePushCordovaPlugin {
     public reportStatus(status: number, label: string, appVersion: string, deploymentKey: string, previousLabelOrAppVersion?: string, previousDeploymentKey?: string) {
        if ((!label || label === previousLabelOrAppVersion) &&
            appVersion === previousLabelOrAppVersion &&
-           deploymentKey == previousDeploymentKey) {
+           deploymentKey === previousDeploymentKey) {
            // No-op since the new appVersion and label is exactly the same as the previous
            // (the app might have been updated via a direct or HockeyApp deployment).
            return;

--- a/www/codePush.ts
+++ b/www/codePush.ts
@@ -63,9 +63,8 @@ class CodePush implements CodePushCordovaPlugin {
      * !!! This function is called from the native side, please make changes accordingly. !!!
      */
     public reportStatus(status: number, label: string, appVersion: string, deploymentKey: string, previousLabelOrAppVersion?: string, previousDeploymentKey?: string) {
-       if ((!label || label === previousLabelOrAppVersion) &&
-           appVersion === previousLabelOrAppVersion &&
-           deploymentKey === previousDeploymentKey) {
+       if (((!label && appVersion === previousLabelOrAppVersion) || label === previousLabelOrAppVersion)
+           && deploymentKey === previousDeploymentKey) {
            // No-op since the new appVersion and label is exactly the same as the previous
            // (the app might have been updated via a direct or HockeyApp deployment).
            return;

--- a/www/codePush.ts
+++ b/www/codePush.ts
@@ -63,6 +63,14 @@ class CodePush implements CodePushCordovaPlugin {
      * !!! This function is called from the native side, please make changes accordingly. !!!
      */
     public reportStatus(status: number, label: string, appVersion: string, deploymentKey: string, previousLabelOrAppVersion?: string, previousDeploymentKey?: string) {
+       if ((!label || label === previousLabelOrAppVersion) &&
+           appVersion === previousLabelOrAppVersion &&
+           deploymentKey == previousDeploymentKey) {
+           // No-op since the new appVersion and label is exactly the same as the previous
+           // (the app might have been updated via a direct or HockeyApp deployment).
+           return;
+       }
+
        var createPackageForReporting = (label: string, appVersion: string): IPackage => {
             return {
                 /* The SDK only reports the label and appVersion.

